### PR TITLE
Add formatting for assertion messages

### PR DESCRIPTION
**Pull request by 0xbe7a**
This PR adds styling codes for assertion messages.
This allows constraints to use styling e.g. to increase readability of their assertion messages. The styling can then be set independently of the platform and converted to ANSI color codes for command line output, for example, and to CSS color tags for HTML reports. The styling tags describe use cases and not concrete colors, so formatter can use arbitrary color palettes and these are not fixed by the constraint. The first use-case which is supported is: numMatch and numDiff for (#89)

The assertion message can be formatted as follows: `[numMatch]3.141[/numMatch][numDifference]6[/numDifference] vs [numMatch]3.141[/numMatch][numDifference]5[/numDifference]`

### CODE REVIEW

File: src/datajudge/constraints/base.py
Comment (ivergara) [inline src/datajudge/constraints/base.py:49]: I like how you are using properties as one of the cookbook intentions, to add behavior to an instance attribute.

File: src/datajudge/formatter.py
Comment (ivergara) [inline src/datajudge/formatter.py:15]: Mostly stylistic, but I'd put the implementation of the `__init__` and `fmt_str` directly in the base class and then let all specific implementations derive from that one instead of the `DefaultFormatter`.

File: src/datajudge/formatter.py
Comment (ivergara) [inline src/datajudge/formatter.py:33]: I don't think you need this here. At least you're not using it in `HtmlFormatter`.

**Review by ivergara — COMMENTED** [review]
Would be great if you add something to the PR body, and also a paragraph somewhere in the documentation about this feature.

Probably also add an entry to the changelog.

**Review by YYYasin19 — APPROVED** [review]
Nice!
~Can you add a little paragraph in the documentation? That way we have a constant reminder that we'll need to extend this in the future :)~

File: src/datajudge/pytest_integration.py
Comment (YYYasin19) [inline src/datajudge/pytest_integration.py:21]: Could you add some tests that validate that this works?
e.g. something that creates a temp test file and calls `pytest` with different options on it.

**Comment by 0xbe7a** [comment]
While i was trying to get html formatting to work, i noticed that there is a new major release 4.0 for pytest coming up. When ansi2html was installed together with pytest-html (the package was not a dependency due to the more permissive software license of ansi2html) pytest-html could already translate the ansi styling to css styling in the caputred log output. Styling in assertion messages was however afaik not supported / bugged. With the rewrite of pytest-html this problem seems to be fixed and pytest-html can now also translate styling in assertion messages itself. This makes at least the html formatter obsolete for us, but I would argue that a separate styling backend is still useful to provide the option for unstyled output for environments without ansi coloring support. I am currently in the process of getting the current 4.0.0 release canidate of pytest-html into conda-forge to allow for testing.

**Comment by ivergara** [comment]
> I am currently in the process of getting the current 4.0.0 release canidate of pytest-html into conda-forge to allow for testing.

You don't have to. Just install from PyPi into your local environment.

**Comment by 0xbe7a** [comment]
> You don't have to. Just install from PyPi into your local environment.

yes, for local testing this works of course. but to work on this pr and test if our integration tests etc. work as well it is handy to be able to target the rc directly in the environment.yml so it can be used directly in the the ci tests too

File: src/datajudge/pytest_integration.py
Comment (0xbe7a) [reply src/datajudge/pytest_integration.py]: I have added an integration test to check if the formatter is selected correctly

File: environment.yml
Comment (0xbe7a) [reply environment.yml:9]: Yes, i think we should wait for the proper release before merging this PR, i just wanted to get it ready for review

Comment (0xbe7a) [reply environment.yml:9]: I like the approach of soft support, but not hard dependence, will add a commit!

Comment (ivergara) [reply environment.yml:9]: > Yes, i think we should wait for the proper release before merging this PR, i just wanted to get it ready for review

That'd be fine if we know the release will be done soonish.

In any case, we should support, as in not break, if a user has a pytest-html < 4 version installed. In that case the html then doesn't have coloring but should break the output there.

**Review by ivergara — APPROVED** [review]

**Pull request merged**
